tests/test_notanorm.py

```diff
@@ -1,182 +1,221 @@
-        db.insert("foo", bar="hi", baz="ho")
 
-        self.assertEqual(db.select("foo", bar="hi")[0].bar, "hi")
+def test_db_upsert(db):
+    db.query("create table foo (bar varchar(32) not null primary key, baz text)")
+    db.insert("foo", bar="hi", baz="ho")
 
-        # updates
-        db.upsert("foo", bar="hi", baz="up")
+    assert db.select("foo", bar="hi")[0].bar == "hi"
 
-        self.assertEqual(db.select("foo", bar="hi")[0].baz, "up")
+    # updates
+    db.upsert("foo", bar="hi", baz="up")
 
-        # inserts
-        db.upsert("foo", bar="lo", baz="down")
+    assert db.select("foo", bar="hi")[0].baz == "up"
 
-        self.assertEqual(db.select("foo", bar="hi")[0].baz, "up")
-        self.assertEqual(db.select("foo", bar="lo")[0].baz, "down")
+    # inserts
+    db.upsert("foo", bar="lo", baz="down")
 
-    def test_db_upsert_non_null(self):
-        db = SqliteDb(self.__fname())
+    assert db.select("foo", bar="hi")[0].baz == "up"
+    assert db.select("foo", bar="lo")[0].baz == "down"
 
-        db.query("create table foo (bar primary key, baz, bop)")
-        db.insert("foo", bar="hi", baz="ho", bop="keep")
 
-        # updates baz... but not bop to none
-        db.upsert_non_null("foo", bar="hi", baz="up", bop=None)
+def test_db_upsert_non_null(db):
+    db.query("create table foo (bar varchar(32) not null primary key, baz text, bop text)")
+    db.insert("foo", bar="hi", baz="ho", bop="keep")
 
-        self.assertEqual(db.select_one("foo").baz, "up")
-        self.assertEqual(db.select_one("foo").bop, "keep")
+    # updates baz... but not bop to none
+    db.upsert_non_null("foo", bar="hi", baz="up", bop=None)
 
-    def test_conn_retry(self):
-        db = SqliteDb(":memory:")
-        db.query("create table foo (x)")
-        db._DbBase__conn_p.close()
+    assert db.select_one("foo").baz == "up"
+    assert db.select_one("foo").bop == "keep"
 
-        with self.assertRaises(sqlite3.ProgrammingError):
-            db.query("create table bar (x)")
 
-        db.retry_errors = (sqlite3.ProgrammingError, )
-        db.query("create table bar (x)")
+def test_model(db):
+    model = DbModel({
+        "foo": DbTable(columns=(
+            DbCol("auto", typ=DbType.INTEGER, autoinc=True, notnull=True),
+            DbCol("blob", typ=DbType.BLOB),
+            DbCol("tex", typ=DbType.TEXT, notnull=True),
+            DbCol("siz3v", typ=DbType.TEXT, size=3, fixed=False),
+            DbCol("siz3", typ=DbType.TEXT, size=3, fixed=True),
+            DbCol("flt", typ=DbType.FLOAT),
+            DbCol("dbl", typ=DbType.DOUBLE),
+        ), indexes=tuple([
+            DbIndex(fields=["auto"], primary=True)
+        ]))
+    })
+    db.create_model(model)
+    check = db.model()
+    assert check == model
 
-    def test_safedb_inmemorydb(self):
-        # test that in memory db's are relatively safe
 
-        db = SqliteDb(":memory:")
+def test_conn_retry(db):
+    db.query("create table foo (x integer)")
```

Comment:
this sneakily tests reconnect, should probably still assert that it fails when reconect is 0